File: src/test/java/tech/sirwellington/alchemy/generator/ChecksTest.java
/*
 * Copyright © 2026. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.generator;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.one;
import static tech.sirwellington.alchemy.generator.Checks.*;
import static tech.sirwellington.alchemy.generator.StringGenerators.alphabeticStrings;

/**
 * Tests for {@link Checks}.
 *
 * @author SirWellington
 */
@DisplayName("Checks should")
class ChecksTest {

    private static final String MESSAGE = "some message";

    @Test
    void testCheckNotNull() {
        Object obj = new Object();
        checkNotNull(obj);
        checkNotNull(obj, MESSAGE);
    }

    @Test
    void testCheckNotNullExpecting() {
        assertThrows(IllegalArgumentException.class, () -> checkNotNull(null));
    }

    @Test
    void testCheckNotNullExpectingWithMessage() {
        assertThrows(
            IllegalArgumentException.class,
            () -> checkNotNull(null, MESSAGE),
            "message should be included"
        );
    }

    @Test
    void testCheckThat() {
        checkThat(true);
        checkThat(true, MESSAGE);
    }

    @Test
    void testCheckThatExpecting() {
        assertThrows(
            IllegalArgumentException.class,
            () -> checkThat(false)
        );
    }

    @Test
    void testCheckThatExpectingWithMessage() {
        assertThrows(
            IllegalArgumentException.class,
            () -> checkThat(false, MESSAGE),
            "message should be included"
        );
    }

    @Test
    void testCheckNotEmptyString() {
        String valid = alphabeticStrings().get();
        checkNotEmpty(valid);

        // null → throws
        assertThrows(
            IllegalArgumentException.class,
            () -> checkNotEmpty((String) null)
        );

        // empty → throws
        assertThrows(IllegalArgumentException.class, () -> checkNotEmpty(""));
    }

    @Test
    void testCheckNotEmptyStringWithMessage() {
        String msg = "test message";
        String valid = one(alphabeticStrings());

        // Valid cases
        checkNotEmpty(valid, msg);
        checkNotEmpty(valid, null);     // uses default message
        checkNotEmpty(valid, "");       // empty string allowed as message

        // Invalid cases
        assertThrows(
            IllegalArgumentException.class,
            () -> checkNotEmpty((String) null, msg)
        );
        assertThrows(
            IllegalArgumentException.class,
            () -> checkNotEmpty("", msg)
        );
    }
}

File: src/test/java/tech/sirwellington/alchemy/generator/ChecksTest.kt
/*
 * Copyright 2015 SirWellington Tech.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.generator

import org.apache.commons.lang3.RandomStringUtils
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.runners.MockitoJUnitRunner
import tech.sirwellington.alchemy.generator.Throwables.assertThrows

/**

 * @author SirWellington
 */
@RunWith(MockitoJUnitRunner::class)
class ChecksTest
{

    private lateinit var message: String

    @Before
    fun setUp()
    {
        message = "some message"
    }

    @Test
    fun testCheckNotNull()
    {
        println("testCheckNotNull")

        val `object` = Any()
        checkNotNull(`object`)
        checkNotNull(`object`, message!!)
    }

    @Test(expected = IllegalArgumentException::class)
    fun testCheckNotNullExpecting()
    {
        println("testCheckNotNullExpecting")

        checkNotNull(null)
    }

    @Test(expected = IllegalArgumentException::class)
    fun testCheckNotNullExpectingWithMessage()
    {
        println("testCheckNotNullExpectingWithMessage")

        checkNotNull(null, message!!)
    }

    @Test
    fun testCheckThat()
    {
        println("testCheckThat")

        checkThat(true)
        checkThat(true, message!!)
    }

    @Test(expected = IllegalArgumentException::class)
    fun testCheckThatExpecting()
    {
        println("testCheckThatExpecting")

        checkThat(false)
    }

    @Test(expected = IllegalArgumentException::class)
    fun testCheckThatExpectingWithMessage()
    {
        println("testCheckThatExpectingWithMessage")

        checkThat(false, message!!)
    }

    @Test
    fun testCheckNotEmptyString()
    {
        println("testCheckNotEmptyString")

        val string = RandomStringUtils.randomAlphanumeric(10)
        checkNotEmpty(string)

        val nullString: String? = null
        assertThrows { checkNotEmpty(nullString) }
                .isInstanceOf(IllegalArgumentException::class.java)

        val emptyString = ""
        assertThrows { checkNotEmpty(emptyString) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testCheckNotEmptyStringWithMessage()
    {
        println("testCheckNotEmptyStringWithMessage")

        val message = RandomStringUtils.randomAlphabetic(100)
        val string = RandomStringUtils.randomAscii(10)
        checkNotEmpty(string, message)
        checkNotEmpty(string, null)
        checkNotEmpty(string, "")

        val nullString: String? = null
        assertThrows { checkNotEmpty(nullString, message) }
                .isInstanceOf(IllegalArgumentException::class.java)

        val emptyString = ""
        assertThrows { checkNotEmpty(emptyString, message) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }
}
